### PR TITLE
nacos  discover support https

### PR DESCRIPTION
**Pull request by huawenyao**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

XXXXX

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java [hunk 1]
```diff
@@ -40,10 +42,7 @@
 import java.net.HttpURLConnection;
 import java.net.SocketTimeoutException;
 import java.net.URL;
-import java.util.ArrayList;
-import java.util.Date;
-import java.util.List;
-import java.util.Properties;
+import java.util.*;
```
Comment (chuntaojun) [inline client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java:45]: don't use import *

File: client/pom.xml [hunk 2]
```diff
@@ -129,8 +129,8 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <configuration>
-                    <source>6</source>
-                    <target>6</target>
+                    <source>8</source>
+                    <target>8</target>
```
Comment (KomachiSion) [inline client/pom.xml:133]: We need to consider about low-level java users.

File: client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java [hunk 3]
```diff
@@ -233,10 +235,19 @@ public HttpResult httpDelete(String path, List<String> headers, List<String> par
...
-                serverListMgr.getContentPath() : "/" + serverListMgr.getContentPath();
-        return serverAddr + contextPath + relativePath;
+            serverListMgr.getContentPath() : "/" + serverListMgr.getContentPath();
+        return StringUtils.removeEnd(serverAddr,"/")+"/"+StringUtils.removeStart(contextPath,"/")+ relativePath;
+    }
+    private String getUrl(String serverAddr, String relativePath,boolean isSSL) {
+        String contextPath = serverListMgr.getContentPath().startsWith("/") ?
+            serverListMgr.getContentPath() : "/" + serverListMgr.getContentPath();
+        String url= StringUtils.removeEnd(serverAddr,"/")+"/"+StringUtils.removeStart(contextPath,"/")+ relativePath;
+        if (isSSL &&url.startsWith(httpPre)){
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java:245]: Here, notice the code formate,  "&& url.startsWith(httpPre)" is perfered.

File: api/src/main/java/com/alibaba/nacos/api/common/Constants.java [hunk 4]
```diff
@@ -23,6 +23,7 @@
  * @author Nacos
  */
 public class Constants {
+    public static final String NACOS_TOEKN_PREFIX="nacos_";
```
Comment (zongtanghu) [inline api/src/main/java/com/alibaba/nacos/api/common/Constants.java:26]: Code formate issue, "NACOS_TOEKN_PREFIX = "nacos_";" is perfered.

File: client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java [hunk 5]
```diff
@@ -59,6 +58,11 @@
 
     private String namespaceId;
 
+    private String  httpPre="http://";
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java:61]: As above.

File: client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java [hunk 6]
```diff
@@ -59,6 +58,11 @@
...
+
+    private String  httpsPre="https://";
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/config/http/ServerHttpAgent.java:63]: As above.

File: client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java [hunk 7]
```diff
@@ -112,11 +118,24 @@ public boolean login(String server) {
             Map<String, String> params = new HashMap<String, String>(2);
             params.put("username", username);
             String body = "password=" + password;
-            String url = "http://" + server + contextPath + LOGIN_URL;
-
-            if (server.contains(Constants.HTTP_PREFIX)) {
-                url = server + contextPath + LOGIN_URL;
+            String url=server;
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java:121]: As above, code formate issue.

File: client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java [hunk 8]
```diff
@@ -131,6 +150,9 @@ public boolean login(String server) {
                 accessToken = obj.getString(Constants.ACCESS_TOKEN);
                 tokenTtl = obj.getIntValue(Constants.TOKEN_TTL);
                 tokenRefreshWindow = tokenTtl / 10;
+                System.setProperty(Constants.NACOS_TOEKN_PREFIX+Constants.ACCESS_TOKEN,accessToken);
+            }else{
+                SECURITY_LOGGER.warn("login  got   accessToken  is empty");
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java:155]: "login got accessToken is empty" is perfered.

File: client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java [hunk 9]
```diff
@@ -112,11 +118,24 @@ public boolean login(String server) {
...
+                    url = "https://" + server + contextPath + LOGIN_URL;
+                } else {
+                    url = "http://" + server + contextPath + LOGIN_URL;
+                }
+            }else{
+                if (HttpClient.ENABLE_HTTPS && !server.startsWith(Constants.HTTPS_PREFIX)) {
+                    url =Constants.HTTPS_PREFIX +StringUtils.removeStart(server,Constants.HTTP_PREFIX)+ contextPath + LOGIN_URL;
+                } else {
+                    url = server + contextPath + LOGIN_URL;
+                }
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java:133]: Here, there are some String character processing, you may use some more optimized methodsm, such as StringBuilder.appened()?

File: client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java [hunk 10]
```diff
@@ -54,6 +55,10 @@
      * User's password
      */
     private String password;
+    /**
+     * NacosServer配置地址
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/security/SecurityProxy.java:59]: Please use English comments.

File: client/src/main/java/com/alibaba/nacos/client/naming/net/HttpClient.java [hunk 11]
```diff
@@ -16,18 +16,26 @@
 package com.alibaba.nacos.client.naming.net;
 
 import com.alibaba.nacos.api.common.Constants;
+import com.alibaba.nacos.client.config.impl.HttpSimpleClient;
 import com.alibaba.nacos.common.utils.HttpMethod;
 import com.alibaba.nacos.common.utils.IoUtils;
 import com.google.common.net.HttpHeaders;
 import org.apache.commons.lang3.StringUtils;
```
Comment (zongtanghu) [inline client/src/main/java/com/alibaba/nacos/client/naming/net/HttpClient.java:23]: In nacos client, I have already removed the dependency of apache-common-lang3, you can merge the code from the develop branch, thanks.

**Comment by zongtanghu** [comment]
Please resolve the conflict codes, firstly, thanks.@huawenyao

**Comment by KomachiSion** [comment]
There are conflict and no response from auth for a long time. I will close this PR. If you want to contribute again, welcome open a new PR.

**Pull request closed without merging**